### PR TITLE
Rework large scale docking

### LINKED ISSUE #507: Rework Large Scale docking using Schematised components
We should rework the large scale docking pipeline to use our new schematised components

### PR DESCRIPTION
**Pull request by hmacdope**
Fixes #507 

Draft  mostly working MVP of large scale docking rework, along with click CLI. Pending changes to docking in #461 but thought I would put this up to share design principles.

### CODE REVIEW

**Comment by hmacdope** [comment]
@ijpulidos @apayne97 @kaminow @jthorton this is probably ready for a first go-over, still a few kinks to work out but very close.

File: asapdiscovery-docking/asapdiscovery/docking/scorer_v2.py [hunk 1]
```diff
@@ -252,6 +271,7 @@ def _score(self, inputs: list[DockingResult]) -> list[Score]:
         results = []
         for inp in inputs:
             schnet_score = self.inference_cls.predict_from_oemol(inp.to_posed_oemol())
+            # save_openeye_pdb(inp.to_posed_oemol(), inp.input_pair.complex.target.target_name + "_+_" + inp.posed_ligand.compound_name + ".pdb")
```
Comment (kaminow) [inline asapdiscovery-docking/asapdiscovery/docking/scorer_v2.py:274]:
```suggestion
```

**Review by kaminow — CHANGES_REQUESTED** [review]
some changes, but this is really nice! I love the direction this is headed and how compact/legible everything is

File: asapdiscovery-data/asapdiscovery/data/schema_v2/fragalysis.py [hunk 2]
```diff
@@ -47,6 +49,90 @@ def data_equal(self, other: FragalysisFactory):
...
+        if (self.xtal_col not in df.columns) or (self.compound_col not in df.columns):
+            raise ValueError(
+                "metadata.csv file must contain a crystal name column and a "
+                "compound name column."
+            )
+
+        try:
+            all_xtal_dirs = os.listdir(self.parent_dir / "aligned")
+        except FileNotFoundError as e:
+            raise FileNotFoundError("No aligned/ directory found in parent_dir.") from e
```
[the remark this answers is not in the document]
Comment (hmacdope) [reply asapdiscovery-data/asapdiscovery/data/schema_v2/fragalysis.py:70]: I think its okay here, as will need to construct the glob here anyway (ie used below.)

File: asapdiscovery-data/asapdiscovery/data/schema_v2/ligand.py [hunk 3]
```diff
@@ -312,3 +316,15 @@ def compound_names_unique(ligands: list[Ligand]) -> bool:
...
+def write_ligands_to_multi_sdf(sdf_name: Union[str, Path], ligands: list[Ligand]):
+    """
+    Dumb way to do this, but just write out each ligand to the same.
+    Alternate way would be to flush each to OEMol and then write out
+    using OE but seems convoluted.
+    """
+    if not sdf_name.split(".")[-1] == "sdf":
+        raise ValueError("SDF name must end in .sdf")
+    for ligand in ligands:
+        ligand.to_sdf(sdf_name, allow_append=True)
```
[the remark this answers is not in the document]
Comment (hmacdope) [reply asapdiscovery-data/asapdiscovery/data/schema_v2/ligand.py:330]: guardrails added.

File: asapdiscovery-data/asapdiscovery/data/schema_v2/structure_dir.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+import dask
+from asapdiscovery.data.dask_utils import actualise_dask_delayed_iterable
+from asapdiscovery.data.schema_v2.complex import Complex
+from asapdiscovery.data.schema_v2.schema_base import DataModelAbstractBase
+from pydantic import BaseModel, Field, validator
+
+
+class StructureDirFactory(BaseModel):
+    parent_dir: Path = Field(
+        description="Top level directory of the Fragalysis database."
```
[the remark this answers is not in the document]
Comment (hmacdope) [reply asapdiscovery-data/asapdiscovery/data/schema_v2/structure_dir.py:15]: Ah sorry, the fragalysis bit is a docstring I messed up up, this is meant to be a a loader for a directroy full of arbitrary PDB files, I have fixed docstring.

File: asapdiscovery-data/asapdiscovery/data/schema_v2/structure_dir.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+        if not v.exists():
+            raise ValueError("parent_dir does not exist.")
+        return v
+
+    @classmethod
+    def from_dir(cls, parent_dir: Path | str):
+        return cls(parent_dir=Path(parent_dir))
+
+    def load(self, use_dask=True, dask_client=None):
+        pdb_files = list(self.parent_dir.glob("*.pdb"))
```
[the remark this answers is not in the document]
Comment (hmacdope) [reply asapdiscovery-data/asapdiscovery/data/schema_v2/structure_dir.py:29]: see above

File: asapdiscovery-data/asapdiscovery/data/selectors/mcs_selector.py [hunk 6]
```diff
@@ -114,12 +114,13 @@ def _select(
             for complex in complexes:
                 complex_mol = complex.ligand.to_oemol()
                 # MCS search
-                mcs = next(iter(mcss.Match(complex_mol, True)))
-                sort_args.append((mcs.NumBonds(), mcs.NumAtoms()))
-
+                try:
+                    mcs = next(iter(mcss.Match(complex_mol, True)))
```
[the remark this answers is not in the document]
Comment (hmacdope) [reply asapdiscovery-data/asapdiscovery/data/selectors/mcs_selector.py:118]: Issue made #545

File: asapdiscovery-docking/asapdiscovery/docking/cli.py [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+    is_flag=True,
+    default=True,
+    help="Whether to write the final docked poses to an SDF file.",
+)
+@click.option(
+    "--use-dask",
+    is_flag=True,
+    default=True,
+    help="Whether to use dask for parallelism.",
+)
```
[the remark this answers is not in the document]
Comment (hmacdope) [reply asapdiscovery-docking/asapdiscovery/docking/cli.py]: Yes, good spot. FIxed.

File: asapdiscovery-data/asapdiscovery/data/schema_v2/ligand.py [hunk 8]
```diff
@@ -251,7 +251,7 @@ def from_sdf(
         oemol = load_openeye_sdf(sdf_file)
         return cls.from_oemol(oemol, **kwargs)
 
-    def to_sdf(self, filename: Union[str, Path]) -> None:
+    def to_sdf(self, filename: Union[str, Path], allow_append=False) -> None:
```
Comment (jthorton) [inline asapdiscovery-data/asapdiscovery/data/schema_v2/ligand.py:254]: Just need to update the doc string for `allow_append`.

File: asapdiscovery-data/asapdiscovery/data/schema_v2/ligand.py [hunk 9]
```diff
@@ -312,3 +316,27 @@ def compound_names_unique(ligands: list[Ligand]) -> bool:
     """
     compound_names = [ligand.compound_name for ligand in ligands]
     return len(set(compound_names)) == len(compound_names)
+
+
+def write_ligands_to_multi_sdf(
```
Comment (jthorton) [inline asapdiscovery-data/asapdiscovery/data/schema_v2/ligand.py:321]: Nice I think we will end up with a few of these types of functions which deal with collections of molecules such as deduplication and loading sets of ligands etc maybe it should have its own file like `ligand_utils` or just `utils`?

File: asapdiscovery-data/asapdiscovery/data/schema_v2/structure_dir.py [hunk 10]
```diff
@@ -0,0 +1,43 @@
+from pathlib import Path
+from typing import List  # noqa: F401
+
+import dask
+from asapdiscovery.data.dask_utils import actualise_dask_delayed_iterable
+from asapdiscovery.data.schema_v2.complex import Complex
+from pydantic import BaseModel, Field, validator
+
+
+class StructureDirFactory(BaseModel):
```
Comment (jthorton) [inline asapdiscovery-data/asapdiscovery/data/schema_v2/structure_dir.py:10]: What does this class do? can we have a doc string description?

File: asapdiscovery-docking/asapdiscovery/docking/workflows/large_scale_docking.py [hunk 11]
```diff
@@ -0,0 +1,359 @@
...
+
+    n_select: PositiveInt = Field(
+        10, description="Number of targets to dock each ligand against, sorted by MCS"
+    )
+
+    top_n: PositiveInt = Field(
+        500, description="Number of docking results to return, ordered by docking score"
+    )
+
+    posit_confidence_cutoff: PositiveFloat = Field(
```
Comment (jthorton) [inline asapdiscovery-docking/asapdiscovery/docking/workflows/large_scale_docking.py:94]: We can use the less-than validator built into field here https://docs.pydantic.dev/1.10/usage/schema/#field-customization

File: asapdiscovery-docking/asapdiscovery/docking/workflows/large_scale_docking.py [hunk 12]
```diff
@@ -0,0 +1,359 @@
...
+    output_dir = inputs.output_dir
+    if output_dir.exists():
+        rmtree(output_dir)
+    output_dir.mkdir()
+
+    logger = FileLogger(
+        inputs.logname, path=output_dir, stdout=True, level=inputs.loglevel
+    ).getLogger()
+
+    logger.info(f"Running large scale docking with inputs: {inputs}")
```
Comment (jthorton) [inline asapdiscovery-docking/asapdiscovery/docking/workflows/large_scale_docking.py:180]: We could also dump the JSON of the inputs to the output folder that will make it easy to run again by just loading the inputs from that file.

File: asapdiscovery-data/asapdiscovery/data/selectors/mcs_selector.py [hunk 13]
```diff
@@ -114,12 +114,13 @@ def _select(
...
+                    sort_args.append((mcs.NumBonds(), mcs.NumAtoms()))
+                except StopIteration:  # no match found
```
Comment (jthorton) [inline asapdiscovery-data/asapdiscovery/data/selectors/mcs_selector.py:120]: Do we have test coverage for this except?

File: asapdiscovery-docking/asapdiscovery/docking/workflows/large_scale_docking.py [hunk 14]
```diff
@@ -0,0 +1,359 @@
...
+
+    # score results
+    logger.info("Scoring docking results")
+    scorer = MetaScorer(
+        scorers=[
+            ChemGauss4Scorer(),
+            GATScorer.from_latest_by_target(inputs.target),
+            SchnetScorer.from_latest_by_target(inputs.target),
+        ]
+    )
```
Comment (jthorton) [inline asapdiscovery-docking/asapdiscovery/docking/workflows/large_scale_docking.py:273]: Do we want to give the user control over the scorers used and how the poses are then ranked?

File: asapdiscovery-docking/asapdiscovery/docking/workflows/large_scale_docking.py [hunk 15]
```diff
@@ -0,0 +1,359 @@
...
+        logger=logger,
+        fail="raise",
+        tag="scores",
+        message="No docking results passed the clash filter",
+    )
+
+    # then order by chemgauss4 score and remove duplicates by ligand id
+    scores_df = scores_df.sort_values(
+        DockingResultCols.DOCKING_SCORE_POSIT.value, ascending=True
+    )
```
Comment (jthorton) [inline asapdiscovery-docking/asapdiscovery/docking/workflows/large_scale_docking.py:318]: Looks like the ML scores are not actually used at all?

Comment (hmacdope) [reply asapdiscovery-docking/asapdiscovery/docking/workflows/large_scale_docking.py:318]: ML scores pushed to postera but not used for ranking at the moment.

File: asapdiscovery-docking/asapdiscovery/docking/workflows/large_scale_docking.py (hunk 14 above)
Comment (hmacdope) [reply asapdiscovery-docking/asapdiscovery/docking/workflows/large_scale_docking.py:273]: Yes to which scorers, for the meantime lets leave the pose ranking as we haven't established a robust protocol for ML ranking.

File: asapdiscovery-docking/asapdiscovery/docking/workflows/large_scale_docking.py (hunk 12 above)
Comment (hmacdope) [reply asapdiscovery-docking/asapdiscovery/docking/workflows/large_scale_docking.py:180]: Yep good idea.

File: asapdiscovery-data/asapdiscovery/data/selectors/mcs_selector.py (hunk 13 above)
Comment (hmacdope) [reply asapdiscovery-data/asapdiscovery/data/selectors/mcs_selector.py:120]: we do now :)

File: asapdiscovery-data/asapdiscovery/data/schema_v2/ligand.py (hunk 9 above)
Comment (hmacdope) [reply asapdiscovery-data/asapdiscovery/data/schema_v2/ligand.py:349]: I might leave it there for now, but can move in future, when more crud arrives.

**Comment by hmacdope** [comment]
@jthorton I think I have addressed all your points. Could I grab a re-review?

**Review by jthorton — APPROVED** [review]
LGTM!

**Pull request merged**